clients/ruby/lib/samsara_sdk/ring_buffer.rb

```diff
@@ -49,36 +54,71 @@ def push(value)
     # @yieldparam [Array<Object>] data Buffer data.
     # @yieldreturn [Boolean] Result of data processing. True if success, false otherwise.
     #
-    # @return [Array<Object>] All buffer's elements.
+    # @return [Array<Object>] All buffer's elements in FIFO order.
     def flush
-      data, snapshot = take_snapshot
+      data, at_mark = take_snapshot
       success = block_given? ? yield(data) : TRUE
-      delete snapshot if success
+      delete at_mark if success
       data
     end
 
     private
 
-    # Get all actual elements from buffer at a given moment in FIFO order.
-    # Make a snapshot of the current buffer.
+    # Helper-method for calculating position in a circle.
+    #
+    # @return [Integer] position of a given pointer.
+    def calculate_position(pointer)
+      @size.nonzero? ? pointer.remainder(@size) : -1
+    end
+
+    # Get array-position of the element that was last added.
+    # Must be used within a synced mutex.
+    #
+    # @return [Integer] position.
+    def high_position
+      calculate_position @high
+    end
+
+    # Get array-position of the element that was last consumed.
+    # Must be used within a synced mutex.
+    #
+    # @return [Integer] position.
+    def low_position
+      calculate_position @low
+    end
+
+    # Make a snapshot of the current buffer at a given moment in FIFO order.
     #
     # @return [Array<Object>] actual elements in FIFO order,
-    # @return [Array<Object>] snapshot of current buffer.
+    # @return [Integer] High-water-mark that snapshot was taken at.
     def take_snapshot
       @mutex.synchronize do
-        i = @pointer == @size - 1 ? 0 : @pointer + 1
-        return (@buffer[i..-1] + @buffer[0...i]).compact, @buffer.dup
+        i = high_position
+        j = low_position + 1
+
+        data = if empty?
+                 []
+               elsif @low == -1
+                 @buffer.compact
+               elsif @high < @size # array is not filled completely
+                 @buffer[j..i]
+               else
+                 @buffer[j..-1] + @buffer[0..i]
+               end
+
+        return data, @high
```

Comment:
Yes. This one is much simpler for future support.
But initially I refused this approach due to the performance issue.

This is a little benchmark.
```ruby
def time
  (Time.now.to_f.round(3) * 1000).to_i
end

size = 500_000

buffer = Array.new(size) { rand(9999) }

puts 'copy by each:'
a = Array.new
start = time
size.times do |i|
  a << buffer[i]
end
puts time - start

puts 'copy by indexes:'
b = Array.new(size)
start = time
i = size / 2
b = buffer[i..size-1] + buffer[0...i]
puts time - start
```
Ruby's C manipulations with arrays using indices is faster then direct `Enum#each`.
On my machine
For 500_000 entries enum is 50 ms - C is 2 ms
For 1_000_000 entries enum is 90 ms - C is 13 ms

Ifs are for handling different cases in order to preserve FIFO.

We have to decide here: performance vs. clear code.